lib/routes/password.js

```diff
@@ -29,13 +29,239 @@ module.exports = function (
+              return password.wrap(wrapKb)
+            })
+            .then((wrapWrapKb) => {
+              // Reset account, delete all sessions and tokens
+              return db.resetAccount(
+                sessionToken,
+                {
+                  verifyHash,
+                  authSalt,
+                  wrapWrapKb,
+                  verifierVersion: password.version
+                }
+              )
+            })
+            .then((result) => {
+              request.emitMetricsEvent('account.changedPassword', {uid})
+              return result
+            })
+        }
+
+        function notifyAccount() {
+          return P.resolve()
+            .then(() => {
+              if (devicesToNotify) {
+                // Notify the devices that the account has changed.
+                push.notifyPasswordChanged(uid, devicesToNotify)
+              }
+
+              log.notifyAttachedServices('passwordChange', request, {
+                uid: sessionToken.uid,
+                iss: config.domain,
+                generation: sessionToken.verifierSetAt
+              })
+
+              const emails = account.emails
+              const geoData = request.app.geo
+              const {
+                browser: uaBrowser,
+                browserVersion: uaBrowserVersion,
+                os: uaOS,
+                osVersion: uaOSVersion,
+                deviceType: uaDeviceType
+              } = request.app.ua
+
+              return mailer.sendPasswordChangedNotification(emails, sessionToken, {
+                acceptLanguage: request.app.acceptLanguage,
+                ip,
+                location: geoData.location,
+                timeZone: geoData.timeZone,
+                uaBrowser,
+                uaBrowserVersion,
+                uaOS,
+                uaOSVersion,
+                uaDeviceType,
+                uid
+              })
+                .catch(e => {
+                  // If we couldn't email them, no big deal. Log
+                  // and pretend everything worked.
+                  log.trace({
+                    op: 'Password.changeFinish.sendPasswordChangedNotification.error',
+                    error: e
+                  })
+                })
+            })
+        }
+
+        function createSessionToken() {
+          return P.resolve()
+            .then(() => {
+              if (! sessionToken.tokenVerified) {
+                return random.hex(16)
+              }
+            })
+            .then(maybeToken => {
+              const {
+                browser: uaBrowser,
+                browserVersion: uaBrowserVersion,
+                os: uaOS,
+                osVersion: uaOSVersion,
+                deviceType: uaDeviceType,
+                formFactor: uaFormFactor
+              } = request.app.ua
+
+              // Create a sessionToken with the verification status of the current session
+              const sessionTokenOptions = {
```

Comment:
If the originating sessionToken was TOTP verified, should this new sessionToken inherit that state?  I don't honestly know, just thinking out loud.